### DISCUSSION TITLE
Possible to load file (atom:Path) in LV2 plugin?

### DISCUSSION THREAD

**Discussion by 38github** [rerdavies/pipedal#85] [Q&A] [answered]
I have compiled Neural Amp Modeler LV2 (NAM) and after restarting services Pipedal found it and it loads! However, I can not choose to open a file in the web GUI to load .nam files. How can this be done?

Secondly... What an amazing thing this is! I have a Focusrite Scarlett 2i2 Gen2 running 128/3 latency without a single xrun on a Raspberry Pi 3 clone (La Frite).

![pipedal_nam_lv2](https://user-images.githubusercontent.com/18294149/228927333-3dd3c643-5f97-4cc8-8865-a8a772c98554.png)

**Comment by 38github** [comment]
Anyone who can help me with this, please?

**Comment by rerdavies** [comment]
I'm doing some work on a Convolution reverb plugin right now, which involves atom:path parameters. As soon as I get clear of it, I'll make sure it works with Neural Amp Modeler as well. Come back in a couple of weeks, and it should be implemented. It's good to have an example to work against!

Great to hear that PiPedal runs well on your Le Frite!  Are you using the 1GB version?

**Reply by 38github** [reply]
I really look forward to it! I am using the 512MB version of the La Frite and Johan Gunnarsson's Debian Bookworm images (https://sd-card-images.johang.se/boards/libretech_s805x_ac.html).

**Reply by rerdavies** [reply]
Implemented in the new v1.1 release, that just went up.

The Neural Amp modeler plugin had bugs, which have been fixed. I've submitted a pull request to the NAM project, which the author is going to accept. It won't work until the pull request has been accepted. 

If you're impatient, this project is the source of the pull request, and has the change: https://github.com/rerdavies/neural-amp-modeler-lv2

CPU use is hefty. It uses about 55% of CPU on my Raspberry Pi. But it runs stably, with no overruns, for me. Let me know if your Le Frite survives. :-)

**Reply by 38github** [reply]
Amazing! I will compile it probably tomorrow. It is late here and I am a little tired. I do really want to try it though. I will let you know if the La Frite breaths its last breath :)

**Comment by 38github** [comment]
I can't install the deb file in Debian Bookworm because of a dependency that is not available. libflac++6v5 is the one and I can only find libflac++10 but `apt --fix-broken install` wants to remove pipedal.

**Reply by 38github** [reply]
I downloaded libflac++6v5 from Bullseye repository and it installed. I then installed your fork of NAM lv2 and the La Frite is being choked BUT it is still able to breath slightly. I had to use one of my "even less cpu" models and about 80% of one core of the four was used. When setup with 128 samples latency I did not experience xruns.

The Cab IR module though... It has a strange sound. When I load an IR that I know well and is very dry I can hear room reverb and the frequency response does not seem correct. Right now it is, to me at least, not usable. Something strange is up with it

**Reply by 38github** [reply]
What sample rate are the included IRs? I uploaded some IRs that I resampled to 48kHz which NAM and my audio interface run at and it sounds great.

**Reply by rerdavies** [reply]
IRs are resampled to the current sample rate at runtime. It could be that the resampling code is broken. :-/ I'll check it carefully.

There's a fairly broad range of sample rates. The CAB IRs are mostly 44100. The reverb IRs are mostly 96000. The IRs are all resampled to the current sample rate at runtime, in the background, before the convolution FFT are calculated. 

I don't have a lot of experience with IRs. The quality of the included reverb IRs are superbly engineered. I suspect that the Cab IRs are not-so-superbly engineered. I'm constrained by licensing Most IRs either don't declare licenses (which means I can't use them, even though the authors intend them to be free), or have license terms that are incompatible with an open-source project.

**Reply by rerdavies** [reply]
@38github The broken libflac++ dependency is annoying. I may be able to fix it by using the C interfaces instead of the C++ interfaces. 

Could you run the following command and tell me what you get:

     apt search libflac |grep libflac

**Reply by 38github** [reply]
libflac++-dev/testing 1.4.2+ds-2 arm64
libflac++10/testing 1.4.2+ds-2 arm64
libflac++6v5/now 1.3.3-2+deb11u1 arm64 [installed,local]
libflac-dev/testing 1.4.2+ds-2 arm64
libflac-doc/testing 1.4.2+ds-2 all
libflac-ocaml/testing 0.3.1-1 arm64
libflac-ocaml-dev/testing 0.3.1-1 arm64
libflac12/testing,now 1.4.2+ds-2 arm64 [installed,automatic]
libflac8/now 1.3.3-2+deb11u1 arm64 [installed,local]

libflac++6v5/now 1.3.3-2+deb11u1 arm64 [installed,local]
and 
libflac8/now 1.3.3-2+deb11u1 arm64 [installed,local]
are the packages I had to get from Debian Bullseye.

**Reply by rerdavies** [reply]
Hmm. Ok. There's no trivial fix for _that_. :-(

The core issue: Raspberry Pi OS is behind the curve on flac decoding libraries.. And your OS is living on the razor's edge. I'm targeting version 8.0/stable of the flac encoder/decoder libraries; you're providing... version 12.0/testing. And ironically, I'm pretty sure I would compile against version 12, even if I can't link to it.

I'm going to have to seek advice on how to deal with issues like that. Versioning on Linux is not something I'm entirely experienced with. In the meantime, at least you're up and running.

**Reply by rerdavies** [reply]
Speaking of which... Do NOT turn on the Wi-Fi Direct. It's based on an ancient network stack used by Raspberry Pi OS  that your distro is almost definitely not using. No idea what will happen, but it certainly will not work.

**Reply by 38github** [reply]
Now I am almost tempted to see what will happen... and report back to you.

**Reply by 38github** [reply]
Nothing crashed or anything.

**Reply by rerdavies** [reply]
Do NOT stick forks in electrical outlets!

:-P

Ok. I guess that's good news. I was worried it would trash your network setup.

**Reply by rerdavies** [reply]
By the way, PiPedal now statically links the FLAC libraries, so the problem with FLAC dependencies should be resolved in the next version. (The change didn't make it into 1.1.21).

**Reply by 38github** [reply]
That is great! I will use Pipedal more during the week and report back if I notice more things that might need reporting.

**Comment by rerdavies** [comment, marked as answer]
Implemented v1.1.
